hwy/base.h

```diff
@@ -980,6 +636,613 @@ using RemoveRef = typename RemoveRefT<T>::type;
+     (HWY_ARCH_X86 && defined(__SSE2__) &&                            \
+      ((HWY_COMPILER_CLANG >= 1600 && !HWY_COMPILER_CLANGCL) ||       \
+       HWY_COMPILER_GCC_ACTUAL >= 1200)))
+#define HWY_HAVE_C11_FLOAT16 1
+#else
+#define HWY_HAVE_C11_FLOAT16 0
+#endif
+
+// If 1, both __bf16 and a limited set of *_bf16 SVE intrinsics are available:
+// create/get/set/dup, ld/st, sel, rev, trn, uzp, zip.
+#if HWY_ARCH_ARM_A64 && !(HWY_COMPILER_CLANG && HWY_COMPILER_CLANG < 1700) && \
+    (defined(__ARM_FEATURE_SVE_BF16) || HWY_COMPILER_GCC_ACTUAL >= 1100)
+#define HWY_ARM_HAVE_BFLOAT16 1
+#else
+#define HWY_ARM_HAVE_BFLOAT16 0
+#endif
+
+#if HWY_ARM_HAVE_BFLOAT16 && \
+    (defined(__ARM_FEATURE_SVE_BF16) || defined(__ARM_FEATURE_SVE))
+#define HWY_SVE_HAVE_BFLOAT16 1
+#else
+#define HWY_SVE_HAVE_BFLOAT16 0
+#endif
+
+#if HWY_ARCH_X86 && defined(__SSE2__) && HWY_COMPILER_GCC_ACTUAL >= 1300
+#define HWY_SSE2_HAVE_BFLOAT16 1
+#else
+#define HWY_SSE2_HAVE_BFLOAT16 0
+#endif
+
+#if HWY_ARM_HAVE_BFLOAT16 || HWY_SSE2_HAVE_BFLOAT16
+#define HWY_HAVE_GCC_OR_ARM_BFLOAT16 1
+#else
+#define HWY_HAVE_GCC_OR_ARM_BFLOAT16 0
+#endif
+
+// Match [u]int##_t naming scheme so rvv-inl.h macros can obtain the type name
+// by concatenating base type and bits. We use a wrapper class instead of a
+// typedef to the native type to ensure that the same symbols, e.g. for VQSort,
+// are generated regardless of F16 support; see #1684.
+struct float16_t {
+#if HWY_NEON_HAVE_FLOAT16C  // ACLE's __fp16
+  using Raw = __fp16;
+#elif HWY_HAVE_C11_FLOAT16                                     // C11 _Float16
+  using Raw = _Float16;
+#elif HWY_CXX_LANG > 202002L && defined(__STDCPP_FLOAT16_T__)  // C++23
+  using Raw = std::float16_t;
+#else
+#define HWY_EMULATE_FLOAT16
+  using Raw = uint16_t;
+  Raw bits;
+#endif  // float16_t
+
+// When backed by a native type, ensure the wrapper behaves like the native
+// type by forwarding all operators. Unfortunately it seems difficult to reuse
+// this code in a base class, so we repeat it in bfloat16_t.
+#ifndef HWY_EMULATE_FLOAT16
+  Raw raw;
+
+  float16_t() noexcept = default;
+  constexpr float16_t(const float16_t&) noexcept = default;
+  constexpr float16_t(float16_t&&) noexcept = default;
+
+  template <typename T, hwy::EnableIf<!IsSame<RemoveCvRef<T>, float16_t>() &&
+                                      IsConvertible<T, Raw>()>* = nullptr>
+  constexpr float16_t(T&& arg) noexcept(
+      noexcept(static_cast<Raw>(DeclVal<T>())))
+      : raw(static_cast<Raw>(static_cast<T&&>(arg))) {}
+
+  template <typename T, hwy::EnableIf<!IsSame<RemoveCvRef<T>, float16_t>() &&
+                                      !IsConvertible<T, Raw>() &&
+                                      IsStaticCastable<T, Raw>()>* = nullptr>
+  explicit constexpr float16_t(T&& arg) noexcept(
+      noexcept(static_cast<Raw>(DeclVal<T>())))
+      : raw(static_cast<Raw>(static_cast<T&&>(arg))) {}
```

Comment:
There is a way to simplify this down to one constructor in C++20 as follows:
```
  template <typename T, hwy::EnableIf<!IsSame<RemoveCvRef<T>, float16_t>() &&
                                      IsStaticCastable<T, Raw>()>* = nullptr>
  explicit(!IsConvertible<T, Raw>()) constexpr float16_t(T&& arg) noexcept(
      noexcept(static_cast<Raw>(DeclVal<T>())))
      : raw(static_cast<Raw>(static_cast<T&&>(arg))) {}
```

The reason to have two separate constructors that differ by only the explicit keyword is to prevent values from being implicitly convertible to hwy::float16_t unless the value is implicitly convertible to hwy::float16_t::Raw and to allow the hwy::float16_t class to be compilable in C++11/C++14/C++17 mode.

There are a few Google Highway targets such as AVX3_SPR/PPC10/RVV/SVE2 that require a C++ compiler that has C++20 support, but most of the other Google Highway targets are compilable with an older C++ compiler that does not have C++20 support.

Reply:
I see, thank you. Keeping both to support earlier C++ versions sounds good.